subsys/net/l2/ethernet/ptp/ptp_master.c

```diff
@@ -0,0 +1,400 @@
+/*
+ * Copyright (c) 2022 Carbon Robotics.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(ptp_master, CONFIG_NET_PTP_LOG_LEVEL);
+
+#include <zephyr/drivers/ptp_clock.h>
+#include <zephyr/net/ethernet.h>
+#include <zephyr/net/net_if.h>
+#include <zephyr/net/net_pkt.h>
+#include <zephyr/net/ptp.h>
+#include <stdlib.h>
+#include <string.h>
+#include <zephyr.h>
+
+#define NET_PTP_MASTER_STACK_SIZE 1024
+
+static const struct net_eth_addr ptp_mcast_eth_addr = { { 0x01, 0x1b, 0x19, 0x00, 0x00, 0x00 } };
+
+struct ptp_master_state {
+	struct net_if *iface;
+	const struct device *clk;
+	struct ptp_port_identity port_identity;
+	struct net_context *context;
+	struct net_pkt *sync_pkt;
+	struct net_if_timestamp_cb sync_sent_cb;
+	uint16_t announce_seq_id;
+	uint16_t sync_seq_id;
+	uint8_t domain_number;
+	uint8_t log_announce_interval;
+	uint8_t log_sync_interval;
+	uint8_t log_min_delay_req_interval;
+	bool running;
+};
+
+static struct ptp_master_state state;
+
+K_KERNEL_STACK_DEFINE(ptp_announce_sender_thread_stack, NET_PTP_MASTER_STACK_SIZE);
+static struct k_thread ptp_announce_sender_thread;
+
+K_KERNEL_STACK_DEFINE(ptp_sync_sender_thread_stack, NET_PTP_MASTER_STACK_SIZE);
+static struct k_thread ptp_sync_sender_thread;
+
+static void handle_ptp_sync_sent(struct net_pkt *pkt);
+static void send_ptp_follow_up(struct net_ptp_time *ts);
+
+static void send_ptp_announce(void)
+{
+	/* prepare body */
+	struct ptp_announce_body ptp_body;
+
+	memset(&ptp_body, 0, sizeof(ptp_body));
+	ptp_body.grandmaster_clock_quality.clock_class = 248;
+	ptp_body.grandmaster_clock_quality.clock_accuracy = 0xfe;
+	ptp_body.grandmaster_clock_quality.offset_scaled_log_var = 0xffff;
+	memcpy(&ptp_body.grandmaster_identity, &state.port_identity.clock_identity,
+		sizeof(ptp_body.grandmaster_identity));
+	ptp_body.time_source = 0xa0;
+
+	/* prepare header */
+	struct ptp_header ptp_hdr;
+
+	memset(&ptp_hdr, 0, sizeof(ptp_hdr));
+	ptp_hdr.message_type = PTP_MESSAGE_TYPE_ANNOUNCE;
+	ptp_hdr.version_ptp = 2;
+	ptp_hdr.minor_version_ptp = 0;
+	ptp_hdr.message_length = sizeof(ptp_hdr) + sizeof(ptp_body);
+	ptp_hdr.domain_number = state.domain_number;
+	ptp_hdr.flag_field = PTP_FLAG_PTP_TIMESCALE;
+	memcpy(&ptp_hdr.source_port_identity, &state.port_identity,
+		sizeof(ptp_hdr.source_port_identity));
+	ptp_hdr.sequence_id = ++state.announce_seq_id;
+	ptp_hdr.log_message_interval = state.log_announce_interval;
+
+	/* allocate packet with PTP settings */
+	struct net_pkt *pkt = net_pkt_alloc_with_buffer(state.iface, ptp_hdr.message_length,
+		AF_UNSPEC, 0, K_NO_WAIT);
```

Comment:
You should verify if the allocation was successfull (`pkt != NULL`) before proceeding, here and in other functions as well.